File: problems/dynamic_programming/knapsack_problem/js/knapsack_problem.js
export const knapsackProblem = (() => {
  /**
   * Builds the DP matrix for the knapsack problem
   * @param {number[][]} items - Arrray of items where each item is [value, weight]
   * @param {number} capacity - The maximum weight the knapsack can carry
   * @returns {number[][]} - DP matrix
   * */
  const fillDpMatrix = (items, capacity) => {
    const n = items.length;
    const dp = Array.from({ length: n + 1 }, () => Array(capacity + 1).fill(0));

    for (let i = 1; i <= n; i++) {
      const currentItemValue = items[i - 1][0];
      const currentItemWeight = items[i - 1][1];

      for (let c = 0; c <= capacity; c++) {
        if (currentItemWeight > c) {
          dp[i][c] = dp[i - 1][c]; // cannot include the current item
        } else {
          dp[i][c] = Math.max(
            dp[i - 1][c], // exclude the item
            dp[i - 1][c - currentItemWeight] + currentItemValue, // include the item
          );
        }
      }
    }
    return dp;
  };

  /**
   * Reconstructs the selected items based on the DP matrix
   * @param {number[][]} dp - DP matrix
   * @param {number[][]} items - Array of items where each item is [value, weight]
   * @param {number} capacity - Maximum weight the knapsack can carry
   * @returns {number[]} - Indices of selected items
   */
  const getSelectedItems = (dp, items, capacity) => {
    const selectedItems = [];
    let n = items.length;
    let c = capacity;

    while (n > 0 && c > 0) {
      if (dp[n][c] !== dp[n - 1][c]) {
        // the item was included in the optimal solution
        selectedItems.push(n - 1);
        c -= items[n - 1][1]; // reduce the capacity
      }
      n--; // move to the previous item
    }
    return selectedItems;
  };

  /**
   * Solves the knapsack problem
   * @param {number[][]} items - Array of items where each item is [value, weight]
   * @param {number} capacity - Maximum weight the knapsack can carry
   * @returns {number[][]} - Result in the format [[totalValue], [selectedItemIndices]]
   */
  return (items, capacity) => {
    const dp = fillDpMatrix(items, capacity);
    const selectedItems = getSelectedItems(dp, items, capacity);

    return [[dp[items.length][capacity], selectedItems]];
  };
})();

File: problems/dynamic_programming/knapsack_problem/python/knapsack_problem.py
def fill_dp_matrix(items: list[list[int]], capacity: int) -> list[list[int]]:
    n = len(items)
    dp: list[list[int]] = [[0] * (capacity + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        current_value = items[i - 1][0]
        current_weight = items[i - 1][1]

        for c in range(capacity + 1):
            if current_weight > c:
                # cannot include the current item
                dp[i][c] = dp[i - 1][c]
            else:
                # include or exclude the current item
                dp[i][c] = max(
                    dp[i - 1][c], dp[i - 1][c - current_weight] + current_value
                )

    return dp


def get_selected_items(
    dp: list[list[int]], items: list[list[int]], capacity: int
) -> list[int]:
    selected_items: list[int] = []
    n = len(items)
    c = capacity

    while n > 0 and c > 0:
        if dp[n][c] != dp[n - 1][c]:
            # the item was included
            selected_items.append(n - 1)
            c -= items[n - 1][1]  # reduce the capacity
        n -= 1  # move to the previous item

    return selected_items


def knapsack_problem(items: list[list[int]], capacity: int) -> list[list[int]]:
    dp = fill_dp_matrix(items, capacity)
    selected_items = get_selected_items(dp, items, capacity)

    return [dp[len(items)][capacity], selected_items]
